File: src/python/problems/next_permutation_test.py
#https://leetcode.com/problems/next-permutation/

from core.problem_base import *

class NextPermutation(ProblemBase):
    def Solution(self, nums: list[int]) -> list[int]:
        """
        Do not return anything, modify nums in-place instead.
        """
        right = len(nums) - 1
        left = right - 1
        found = False

        while left >= 0:
            if not found and nums[left] < nums[right]:
                found = True
                right = len(nums) - 1
            elif not found:
                left -= 1
                right -= 1
            else:
                if nums[left] < nums[right]:
                    nums[left], nums[right] = nums[right], nums[left]
                    break
                right -= 1

        nums[left + 1:] = reversed(nums[left + 1:])


if __name__ == '__main__':
    TestGen(NextPermutation) \
        .Add(lambda tc: tc.Param("nums", [1,2,3]).Result([1,3,2])) \
        .Add(lambda tc: tc.Param("nums", [3,2,1]).Result([1,5,1])) \
        .Add(lambda tc: tc.Param("nums", [1,1,5]).Result([1,5,1])) \
        .Add(lambda tc: tc.Param("nums", [1,3,2]).Result([2,1,3])) \
        .Run()

File: src/challenges/go/problems/next_permutation_test.go
// https://leetcode.com/problems/next-permutation/

package problems

import (
	"testing"
	"github.com/dobermanch/leetcode/core"
)

type NextPermutation struct{}

func TestNextPermutation(t *testing.T) {
	gen := core.TestSuite[NextPermutation]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([]int{1,2,3}).Result([]int{1,3,2})
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{3,2,1}).Result([]int{1,2,3})
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{1,1,5}).Result([]int{1,5,1})
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{1,3,2}).Result([]int{2,1,3})
	}).Run(t)
}

func (NextPermutation) Solution(nums []int) []int {
	right := len(nums) - 1
	left := right - 1
	found := false

	for (left >= 0) {
		if (!found && nums[left] < nums[right]) {
			found = true
			right = len(nums) - 1
		} else if (!found) {
			left--
			right--
		} else {
				if (nums[left] < nums[right]) {
				nums[left], nums[right] = nums[right], nums[left]
				break
			}
			right--
		}
	}

	for i, j := left + 1, len(nums) - 1; i < j; i, j = i + 1, j - 1 {
		nums[i], nums[j] = nums[j], nums[i]
	}

	return nums
}
